### PR TITLE
Fixed assets investments note API

### PR DESCRIPTION
**Pull request by JMetherell**
Adds POST/PUT/GET/DELETE, controller, service, transformer and models for the fixed assets investments note. Does not include validation.

Resolves:
SFA-1112

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+        setMetadataOnRestObject(rest, transaction, companyAccountId);
+
+        FixedAssetsInvestmentsEntity entity = transformer.transform(rest);
+        entity.setId(generateID(companyAccountId));
+
+        try {
+            repository.save(entity);
+        } catch (MongoException me) {
+            DataException dataException = new DataException("Failed to update" +
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java]: Very minor, but you'll need a space after 'update':
```
"Failed to update " + 
```

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/FixedAssetsInvestmentsValidator.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import javax.validation.Valid;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.api.accounts.exception.DataException;
+import uk.gov.companieshouse.api.accounts.model.rest.notes.fixedassetsinvestments.FixedAssetsInvestments;
+import uk.gov.companieshouse.api.accounts.model.validation.Errors;
+import uk.gov.companieshouse.api.accounts.transaction.Transaction;
+
+@Component
+public class FixedAssetsInvestmentsValidator extends BaseValidator {
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/FixedAssetsInvestmentsValidator.java]: Do we have this class if no totalling / cross validation is required?

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            ResponseObject<FixedAssetsInvestments> response = fixedAssetsInvestmentsService
+                    .create(fixedAssetsInvestments, transaction, companyAccountId, request);
+
+            return apiResponseMapper
+                .map(response.getStatus(), response.getData(), response.getErrors());
+
+        } catch (DataException ex) {
+            final Map<String, Object> debugMap = createDebugMap(companyAccountId, transaction,
+                "Failed to update fixedAssetsInvestments resource");
+            LOGGER.errorRequest(request, ex, debugMap);
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java]: The logging in the controller has been updated across all the api controllers. Using a LogHelper. 

`LoggingHelper.logException(companyAccountId, transaction,
                "Failed to retrieve fixedAssetsInvestments resource", ex, request);`

These should be updated to match.

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                .update(fixedAssetsInvestments, transaction, companyAccountId, request);
+
+            return apiResponseMapper
+                .map(response.getStatus(), response.getData(), response.getErrors());
+
+        } catch (DataException ex) {
+
+            final Map<String, Object> debugMap = createDebugMap(companyAccountId, transaction,
+                "Failed to update fixedAssetsInvestments resource");
+            LOGGER.errorRequest(request, ex, debugMap);
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java]: as above

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            ResponseObject<FixedAssetsInvestments> response = fixedAssetsInvestmentsService
+                .findById(fixedAssetsInvestmentsId, request);
+
+            return apiResponseMapper.mapGetResponse(response.getData(), request);
+
+        } catch (DataException de) {
+
+            final Map<String, Object> debugMap = createDebugMap(companyAccountId, transaction,
+                "Failed to retrieve fixedAssetsInvestments resource");
+            LOGGER.errorRequest(request, de, debugMap);
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java]: as above

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        try {
+            ResponseObject<FixedAssetsInvestments> response =
+                    fixedAssetsInvestmentsService.delete(companyAccountsId, request);
+
+            return apiResponseMapper.map(response.getStatus(), response.getData(), response.getErrors());
+        } catch (DataException de) {
+            final Map<String, Object> debugMap = createDebugMap(companyAccountsId, transaction,
+                "Failed to delete fixedAssetsInvestments resource");
+            LOGGER.errorRequest(request, de, debugMap);
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java]: as above

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java (hunk 3 above)
Comment (JMetherell) [reply src/main/java/uk/gov/companieshouse/api/accounts/controller/FixedAssetsInvestmentsController.java]: Ok - will update to match..

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        setMetadataOnRestObject(rest, transaction, companyAccountId);
+
+        FixedAssetsInvestmentsEntity entity = transformer.transform(rest);
+        entity.setId(generateID(companyAccountId));
+
+        try {
+            repository.insert(entity);
+        } catch (DuplicateKeyException e) {
+            LOGGER.errorRequest(request, e, getDebugMap(transaction, companyAccountId,
+                entity.getId()));
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java]: the services have all been updated to no longer LOG the exceptions in the service layer. So can be removed and the logging is handled within the controller.

**Comment by Mathsh** [comment]
You'll need to update the tests after the logging elements have been altered

File: src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java (hunk 7 above)
Comment (JMetherell) [reply src/main/java/uk/gov/companieshouse/api/accounts/service/impl/FixedAssetsInvestmentsService.java]: Services, controller and tests updated to match the tidier logging approach that is in develop

**Pull request merged**